src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs

```diff
@@ -988,26 +986,40 @@ private BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression
                 method: null,
                 constrainedToTypeOpt: null);
 
-            // (tempx.HasValue OP tempy.HasValue)
-            BoundExpression rightExpression = MakeBinaryOperator(
-                syntax: syntax,
-                operatorKind: rightOperator,
-                loweredLeft: callX_HasValue,
-                loweredRight: callY_HasValue,
-                type: boolType,
-                method: null,
-                constrainedToTypeOpt: null);
+            var canOmitHasValueCheck = yNonNull is { ConstantValue.IsDefaultValue: false } || xNonNull is { ConstantValue.IsDefaultValue: false };
 
             // result = (tempx.GetValueOrDefault() OP tempy.GetValueOrDefault()) &
             //          (tempx.HasValue OP tempy.HasValue)
```

Comment:
Unrelated to the changes. @RikkiGibson Do you know why this is a bitwise AND instead of logical AND? Wouldn't that evaluate both sides and be slower? or there are other aspects I'm not aware of?

Reply:
I think because we're just reading two fields here, it's faster to always do both reads than to branch on the value of the left operand.